### PR TITLE
planner: support hint USE_CASCADES(true|false) to use/disuse cascades planner in a single query

### LINKED ISSUE #15215: UCP: support hint 'USE_CASCADES(ON|OFF)' to use/disuse cascades planner in a single query
## Description
Cascades planner is a new optimizer in TiDB. Currently, we uses a session variable `tidb_enable_cascades_planner` to determine whether uses the cascades planner or not. But sometimes, we only want to use cascades planner for a single query, without affecting other queris in the same session. 
For example:
```sql
# This query should be forced to use cascades planner.
select /*+ USE_CASCADES(ON) */ * from t;
```

```sql
# This query should be forced to use core planner(the old one).
select /*+ USE_CASCADES(OFF) */ * from t;
```
## Score

* 300
## Mentor(s)

* @francis0407 

Contact the mentors: **#tidb-challenge-program** channel in [TiDB Community](https://join.slack.com/t/tidbcommunity/shared_invite/enQtNzc0MzI4ODExMDc4LWYwYmIzMjZkYzJiNDUxMmZlN2FiMGJkZjAyMzQ5NGU0NGY0NzI3NTYwMjAyNGQ1N2I2ZjAxNzc1OGUwYWM0NzE)  Slack Workspace

## Recommended Skills

* Golang
* SQL

## Learning Materials
* PR for `tidb_enable_cascades_planner`: https://github.com/pingcap/tidb/pull/7879
* Chinses blog for cascades planner: https://pingcap.com/blog-cn/tidb-cascades-planner/

### PR DESCRIPTION
**Pull request by gauss1314**
UCP #15215 

### What problem does this PR solve?

Close #15215. Add support hint USE_CASCADES(true|false) to use/disuse cascades planner in a single query.

### What is changed and how it works?

Handle USE_CASCADES hint.

### Check List

Tests

 - Unit test

### CODE REVIEW

File: planner/optimize.go [hunk 1]
```diff
@@ -344,6 +347,15 @@ func handleStmtHints(hints []*ast.TableOptimizerHint) (stmtHints stmtctx.StmtHin
 		stmtHints.HasAllowInSubqToJoinAndAggHint = true
 		stmtHints.AllowInSubqToJoinAndAgg = useToJAHint.HintData.(bool)
 	}
+	// Handle USE_CASCADES
+	if useCascadesHintCnt != 0 {
+		if useCascadesHintCnt > 1 {
+			warn := errors.New("There are multiple USE_CASCADES hints, only the last one will take effect")
```
Comment (zz-jason) [inline planner/optimize.go]:
```suggestion
			warn := errors.Errorf("multiple USE_CASCADES() detected, only the last one takes effect: USE_CASCADES(%v)", useCascadesHint.HintData.(bool))
```

File: session/session_test.go [hunk 2]
```diff
@@ -3048,6 +3048,17 @@ func (s *testSessionSuite2) TestStmtHints(c *C) {
...
 
+	// Test USE_CASCADES hint
+	tk.Se.GetSessionVars().SetEnableCascadesPlanner(true)
+	tk.MustExec("select /*+ USE_CASCADES(false) */ 1;")
+	c.Assert(tk.Se.GetSessionVars().GetEnableCascadesPlanner(), IsFalse)
+	tk.Se.GetSessionVars().SetEnableCascadesPlanner(false)
+	tk.MustExec("select /*+ USE_CASCADES(true) */ 1;")
+	c.Assert(tk.Se.GetSessionVars().GetEnableCascadesPlanner(), IsTrue)
+	tk.MustExec("select /*+ USE_CASCADES(false), USE_CASCADES(true) */ 1;")
+	c.Assert(tk.Se.GetSessionVars().StmtCtx.GetWarnings(), HasLen, 1)
```
Comment (zz-jason) [inline session/session_test.go]: please also test the warning message.

Comment (gauss1314) [reply session/session_test.go]: Hi, is it appropriate to write this way?
`c.Assert(tk.Se.GetSessionVars().StmtCtx.GetWarnings()[0].Err.Error(), Equals, "USE_CASCADES() is defined more than once, only the last definition takes effect: USE_CASCADES(true)")`

Comment (zz-jason) [reply session/session_test.go]: exactly.

**Pull request merged**